### PR TITLE
Active Record + PostgreSQL: native support for `timestamp with time zone`

### PR DESCRIPTION
**Pull request by ghiculescu**
In https://github.com/rails/rails/issues/21126 it was suggested to make "timestamp with time zone" the default type for datetime columns in PostgreSQL. This is in line with PostgreSQL [best practices](https://wiki.postgresql.org/wiki/Don't_Do_This#Don.27t_use_timestamp_.28without_time_zone.29). This PR lays some groundwork for that.

This PR adds a configuration option, `ActiveRecord::ConnectionAdapters::PostgreSQLAdapter.datetime_type`. The default is `:timestamp` which preserves current Rails behavior of using "timestamp without time zone" when you do `t.datetime` in a migration. If you change it to `:timestamptz`, you'll get "timestamp with time zone" columns instead.

If you change this setting in an existing app, you should immediately call `bin/rails db:migrate` to ensure your `schema.rb` file remains correct. If you do so, then existing columns will not be impacted, so for example if you have an app with a mixture of both types of columns, and you change the config, schema dumps will continue to output the correct types.

This PR also adds two new types that can be used in migrations: `t.timestamp` and `t.timestamptz`.

```ruby
ActiveRecord::ConnectionAdapters::PostgreSQLAdapter.datetime_type = :timestamp # default value is :timestamp

create_table("foo1") do |t|
  t.datetime :default_format # "timestamp without time zone"
  t.timestamp :without_time_zone # "timestamp without time zone"
  t.timestamptz :with_time_zone # "timestamp with time zone"
end

ActiveRecord::ConnectionAdapters::PostgreSQLAdapter.datetime_type = :timestamptz

create_table("foo2") do |t|
  t.datetime :default_format # "timestamp with time zone" <-- note how this has changed!
  t.timestamp :without_time_zone # "timestamp without time zone"
  t.timestamptz :with_time_zone # "timestamp with time zone"
end

ActiveRecord::ConnectionAdapters::PostgreSQLAdapter::NATIVE_DATABASE_TYPES[:my_custom_type] = { name: "custom_datetime_format_i_invented" }
ActiveRecord::ConnectionAdapters::PostgreSQLAdapter.datetime_type = :my_custom_type

create_table("foo3") do |t|
  t.datetime :default_format # "custom_datetime_format_i_invented"
  t.timestamp :without_time_zone # "timestamp without time zone"
  t.timestamptz :with_time_zone # "timestamp with time zone"
end
```

**Notes**

- This PR doesn't change the default `datetime` format. The default is still "timestamp without time zone". A future PR could do that, but there was enough code here just getting the config option right.
- See also https://github.com/rails/rails/pull/41395 which set some groundwork (and added some tests) for this.
- This reverts some of https://github.com/rails/rails/pull/15184. https://github.com/rails/rails/pull/15184 alluded to issues in XML serialization, but I couldn't find any related tests that this broke.

### CODE REVIEW

**Comment by ghiculescu** [comment]
@rafaelfranca tagging you in since you commented here https://github.com/rails/rails/issues/21126#issuecomment-732463383 (but let me know if I should tag someone else - you get tagged in for everything). How's this looking so far?

I'm not confident about the CI bit https://github.com/rails/buildkite-config/pull/11 and don't really know how to test that it will work. I have been running `POSTGRESQL_DATETIME_TYPE=timestamp bundle exec rake test:postgresql` locally to confirm tests pass. The way that's implemented is [fairly bad](https://github.com/rails/rails/pull/41084/files#diff-2468c670eb10c24bd2823e42708489a336d6f21c6efc7e3c4a574166fa77bb22R143-R145) too, but I couldn't find/think of any better approaches.

**Comment by rafaelfranca** [comment]
I don't think this is going in the direction I commented. I said we should change the database type of the columns generated by `t.timestamps` on migrations to be `timestampz` and other related methods on migrations but we should not change all the currently timestamp columns to be handled like timestampz, neither we should change the datetime type to be timestampz.

**Comment by ghiculescu** [comment]
@rafaelfranca this doesn't change current timestamp columns. It only changes what will be generated by `t.timestamps`, `t.add_column`, etc. in future migrations. That was what I took away from https://github.com/rails/rails/issues/21126 - have I misunderstood?

If you prefer to just have `t.timestamp` create `timestamptz` but other columns still be created as `t.timestamp`, that's possible, but I'm not sure I understand why you'd want to be inconsistent between different columns.

For existing apps I wouldn't even recommend this change (hence the warning in changelog and new framework defaults). But I think it is a better default for new ones.

**Comment by matthewd** [comment]
Sorry I'm a bit out of the loop here, but this feels like something that should go through migration versioning, not framework defaults, to me.

**Comment by ghiculescu** [comment]
I think I figured out where we are talking past each other.

I've always used `structure.sql` schema dumps. So from my perspective once the migration is written and committed, the column type is cemented regardless of changes to `NATIVE_DATABASE_TYPES`. But if you use `schema.rb`, then changes to `NATIVE_DATABASE_TYPES` would cause your schema to change next time you run tests or load a new development environment, which is bad.

**Comment by ghiculescu** [comment]
while investigating ^ I think I found another bug: #41395

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb [hunk 1]
```diff
@@ -785,6 +806,8 @@ def configure_connection
           unless variables["timezone"]
             if ActiveRecord::Base.default_timezone == :utc
               variables["timezone"] = "UTC"
+            elsif ActiveRecord::Base.default_timezone == :local
+              variables["timezone"] = ENV["TZ"]
```
Comment (ghiculescu) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:810]: confirm if still needed

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb [hunk 2]
```diff
@@ -882,20 +905,34 @@ def add_pg_encoders
...
+              PG::TextDecoder::TimestampUtc :
+              PG::TextDecoder::TimestampWithTimeZone
 
-            @timestamp_decoder = decoder_class.new(@timestamp_decoder.to_h)
-            @connection.type_map_for_results.add_coder(@timestamp_decoder)
+            if @timestamp_decoder
+              @timestamp_decoder = timestamp_decoder_class.new(@timestamp_decoder.to_h)
+              @connection.type_map_for_results.add_coder(@timestamp_decoder)
+            end
+            if @timestamptz_decoder
```
Comment (ghiculescu) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:920]: confirm if still needed

**Comment by ghiculescu** [comment]
@matthewd @rafaelfranca I refactored this PR a fair bit thanks to your comments. Have updated the PR body accordingly. Could you please give it another look and let me know if this seems more on the right track?

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb [hunk 3]
```diff
@@ -319,7 +338,9 @@ def discard! # :nodoc:
       end
 
       def native_database_types #:nodoc:
-        NATIVE_DATABASE_TYPES
+        types = NATIVE_DATABASE_TYPES.dup
+        types[:datetime] = types[datetime_type]
+        types
```
Comment (pixeltrix) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:343]: Might be worth caching this - `native_database_types` is mainly (only?) called by SchemaDumper but it does it twice for every column dumped.

Comment (ghiculescu) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Good idea - updated.

Comment (ghiculescu) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Hmmm. The tests [didn't like that](https://buildkite.com/rails/rails/builds/77093#ea12f75d-abe4-4f02-aaf1-710a4735d707), I guess because they aren't reloading the adapter between each test. I'm going to revert for now and investigate if memoization is necessary - I don't think this is called much at runtime.

Comment (pixeltrix) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:343]: > Hmmm. The tests [didn't like that](https://buildkite.com/rails/rails/builds/77093#ea12f75d-abe4-4f02-aaf1-710a4735d707), I guess because they aren't reloading the adapter between each test. I'm going to revert for now and investigate if memoization is necessary - I don't think this is called much at runtime.

Clearing the instance variable used for caching the results before/after the relevant tests should fix that

Comment (ghiculescu) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Yep - that's what I ended up doing. This is what the changes ended up looking like: https://gist.github.com/ghiculescu/eb14ab894824d6914a1a2360d8b14937

It's fairly messy, I'd love any ideas on how to simplify.

Comment (pixeltrix) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: I think the implementation is fine, maybe just extract the method to a module and put it in `test/support` so we're not duplicating it.

Comment (ghiculescu) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Yep that’s what I did in a subsequent commit. It’s now all a single helper method. 

Thanks for your help here! Did you have any other thoughts on this PR?

Comment (pixeltrix) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Is the comment @matthewd made about changing existing columns in schema.rb been addresses - I think so but just want to confirm.

Comment (ghiculescu) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Yes, I think so too. 

Some of the comments are stale since the direction of the PR changed a bit (eg the reference to using new framework defaults). The PR body is up to date though.

Comment (pixeltrix) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: > Yes, I think so too.

Can you confirm by running `rails db:schema:dump` on a PostgreSQL app with timestamp and timestamptz columns - once with `datetime_type` set to `:timestamp` and once with it set to `:timestamptz`. If there's no changes to the schema.rb file then we're good to go.

Comment (pixeltrix) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: You could add a railties test for this in: 
https://github.com/rails/rails/blob/main/railties/test/application/configuration_test.rb

Comment (ghiculescu) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Yep, will do tomorrow and report back 👍

Comment (kamipo) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: > Is the comment @matthewd made about changing existing columns in schema.rb been addressed - I think so but just want to confirm.

What @matthewd said in https://github.com/rails/rails/pull/41084#issuecomment-776414131 is that the configureation should not change the behavior for all existing migrations. See #34970, #25395, #26266 for your reference.

Comment (ghiculescu) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: @pixeltrix  @kamipo thanks for those links. I had a look through them, and I think this PR is a bit different to those. In all those cases, we are changing how Rails handles certain commands in `db:migrate` without configuration.

- https://github.com/rails/rails/pull/34970 makes `t.timestamps` mean `t.timestamps null: false` from 6.0 onwards, but `t.timestamps null: true` before 6.0.
- https://github.com/rails/rails/pull/25395 makes `create_table(id: :uuid)` use a different UUID function from 5.1 onwards.
- https://github.com/rails/rails/pull/26266 makes `create_table` create bigint ID columns, for tables created on 5.1 onwards.

As I understand it, the point of all these is so that if you run an older migration, it still does the same thing. eg. if you run a migration created on Rails 5, it'll still use the old UUID function even if you have upgraded to 5.1.

(Sorry to spell out the obvious to you, but this is helping me confirm my understanding.)

The difference in this PR is that we are adding a new configuration as well. Nothing will change by default; to use it you have to do something like `ActiveRecord::ConnectionAdapters::PostgreSQLAdapter.datetime_type = : timestamptz ` in an initializer.

Given that, I'm not sure there's a way to make it work with https://github.com/rails/rails/blob/main/activerecord/lib/active_record/migration/compatibility.rb because that would require something in either Rails or the user's code to indicate from which migration onward that setting was changed.

To work around this, I have added this warning to the PR body, release notes, and configuration guide:

> If you change this setting in an existing app, you should immediately call `bin/rails db:migrate` to ensure your `schema.rb` file remains correct.

If you do this, then with the above change, columns that were `t.datetime` will become `t.timestamp`, and `t.timestamptz` will become `t.datetime`. I am working on a test to demonstrate this.

Given that, I don't think any changes in https://github.com/rails/rails/blob/main/activerecord/lib/active_record/migration/compatibility.rb are required. Does this make sense, and am I on track?

Comment (pixeltrix) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Isn't the difference is that in older migrations it should ignore the new config and always use `timestamp`?

Comment (ghiculescu) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: ... 🤦 Yeah, I think so. Let me try that out.

Comment (ghiculescu) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: I can remove the comments and I will add tests, but is this what you had in mind? https://github.com/rails/rails/pull/41084/commits/c4290dae3da94b610a070b84377b94d201305465

Comment (pixeltrix) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: @ghiculescu yes, that's right - the inheritance chain is slightly confusing so I had think about it for a moment 🤯

Comment (ghiculescu) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Haha, you're telling me! I think some comments about it for future developers would be helpful. Is there a preferred way to do that? I was thinking of adding a few lines explaining the inheritance chain here: https://github.com/rails/rails/blob/main/activerecord/lib/active_record/migration/compatibility.rb#L16 but I don't want them to end up as public documentation.

File: activerecord/lib/active_record/migration/compatibility.rb [hunk 4]
```diff
@@ -16,6 +16,47 @@ def self.find(version)
       V7_0 = Current
 
       class V6_1 < V7_0
+        class PostgreSQLCompat
+          def self.compatible_timestamp_type(type, connection)
```
Comment (ghiculescu) [inline activerecord/lib/active_record/migration/compatibility.rb:20]: @pixeltrix @kamipo I've implemented the migration versioning here, tests are in `activerecord/test/cases/schema_dumper_test.rb`.

I think it's working correctly and tested for all scenarios:

- [x] Migration created in Rails 6.1, using default value for `datetime_type` -> no changes to schema.rb; `t.datetime` columns will continue to say `t.datetime`. The tests added for this pass without changes on `main`.
- [x] Migration created in Rails 6.1, with `datetime_type` changed to `:timestamptz` in an initializer -> if the migration said `t.datetime`, schema.rb will be changed to say `t.timestamp`. See comments + assertions [here](https://github.com/rails/rails/pull/41084/files#diff-f4545f25496b789b9cde6e484343a566d476078a22b3b770a80f3ec6b989ea64R605-R614) for why this is necessary.
- [x] Migration created in Rails 7+,  using default value for `datetime_type` -> `t.datetime` columns say `t.datetime` in schema.rb. `t.timestamp` columns say `t.datetime` in schema.rb. `t.timestamptz` columns say `t.timestamptz` in schema.rb.
- [x] Migration created in Rails 6.1, with `datetime_type` changed to `:timestamptz` -> `t.datetime` columns say `t.datetime` in schema.rb. `t.timestamp` columns say `t.timestamp` in schema.rb. `t.timestamptz` columns say `t.datetime` in schema.rb.

**Review by pixeltrix — APPROVED** [review]

**Comment by pixeltrix** [comment]
@ghiculescu thanks for your patience and efforts in getting this merged in 👍🏻

**Comment by ghiculescu** [comment]
Thanks so much for your help @pixeltrix !

**Comment by tomichal** [comment]
When might this be released in Rails? Thanks!

**Comment by pixeltrix** [comment]
@tomichal not anytime soon - this is a Rails 7.0 feature so don't expect it to appear in a 6.1 patch release.

**Comment by tomichal** [comment]
Thank you for the prompt response @pixeltrix . Any tips for making the "timestamp with time zone" column type stick in the db/schema.rb until this is rolled out?

**Comment by ghiculescu** [comment]
> Any tips for making the "timestamp with time zone" column type stick in the db/schema.rb until this is rolled out?


@tomichal this PR may help: https://github.com/rails/rails/pull/41395

If you are running [edge Rails](https://shopify.engineering/living-on-the-edge-of-rails), both #41395 and this PR are on the main branch now, so you can use them both. Otherwise they'll both be in Rails 7 :)

**Comment by tomichal** [comment]
Thanks much @ghiculescu 
Just to clarify -- Currently in Rails 6.1.3.2 I can do a migration like this:
`add_column :some_table, :foo_date, :timestamptz`
and it will make the `foo_date` column in Postgres DB be a "timestamp with time zone". However the db/schema.rb will specify the field as `t.datetime :foo_date`,  which translates into a "timestamp without time zone" in Postgres.

**Comment by ghiculescu** [comment]
Yep, that's right. With https://github.com/rails/rails/pull/41395 that will be resolved. For now, we have been using `structure.sql` instead of `schema.rb` because of this - that was some of my motivation for working on that PR.

**Comment by tomichal** [comment]
Thanks for the clarification @ghiculescu

**Comment by Dearest** [comment]
Because of this PR, the behavior of rails7 and rails6 is not consistent when dealing with timestamptz.

For example: in rails6, created_at field type is timestamptz. then `created_at.class` will be `ActiveSupport::TimeWithZone`. But the same database, same field, in rails7 `created_at.class` is `Time`.

The behavior is inconsistent, how can I solve this problem?

**Comment by ghiculescu** [comment]
@Dearest that sounds like a bug, but it's not something I've seen. Could you provide an executable test case to help me replicate? https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#create-an-executable-test-case

**Comment by Dearest** [comment]
@ghiculescu 
Using this testcase, I can't reproduce the problem I'm having.
```ruby
# frozen_string_literal: true

require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  # gem 'rails', '7.0.2.2'
  gem 'rails', '6.1.4.1'
  gem 'pg'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# require "active_record/connection_adapters/postgresql_adapter"
# ActiveRecord::ConnectionAdapters::PostgreSQLAdapter.datetime_type = :timestamptz

# You will need actual Postgres to do this,
ActiveRecord::Base.establish_connection(adapter: 'postgresql', database: 'dummy_test', host: '127.0.0.1',
                                        username: 'postgres', password: '5432')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.datetime :happened_at, null: true

    t.timestamps
  end
end

class TestApp < Rails::Application
  config.root = __dir__
  config.time_zone = 'Beijing'
  Rails.application.initialize!
end

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def setup
    Post.destroy_all
  end

  def test_happened_at_class_name
    Post.create(happened_at: Time.zone.now)
    assert_equal ActiveSupport::TimeWithZone, Post.first.happened_at.class
  end
end

```

With this test case, `created_at.class` is always `Time`, 
but I created a brand new project using `rails 6.1.4.2` with the following code and `created_at.class` is `ActiveSupport::TimeWithZone`.

```ruby
class CreatePosts < ActiveRecord::Migration[6.1]
  def change
    create_table :posts do |t|
      t.datetime :happened
      t.timestamps
    end
  end
end
```
```ruby
class Post < ApplicationRecord
end
```
image

I don't know what's going wrong

**Comment by ghiculescu** [comment]
Thanks. I’m offline for a few days but I’ll take a look next week 👍

**Comment by Dearest** [comment]
@ghiculescu 
I'll elaborate on the problem I'm having and I hope you can help me.

I am trying to upgrade Rails 6.x to Rails 7.x.

I have a project that connects to multiple databases. In these databases some time fields are of type `timestamp` and some fields are `timestamptz`. This means that I have both timestamp and timestamptz in the databases I am connected to.
Now I have a model called `Post`, which has a `published_at` attribute, and this field is of type timestamptz in the database (this column is use sql create, not rails migration.)

```ruby
# Rails6.1.4.2
Post.last.published_at.class
# ==> ActiveSupport:: TimeWithZone
Post.last.published_at.to_s
# ==> 2021-12-08 18:13:34
```
everything is fine
When I upgraded to Rails 7. same model, same fields

```ruby
# Rails 7.0.1
Post.last.published_at.class
# ==> Time
Post.last.published_at.to_s
# ==> 2022-02-23 02:33:36 UTC
```
(Please don't care about the difference in values, what I'm trying to convey is that the Class is different, as well as the return value format I expect for to_s is different)

To summarize: Rails 6.x formats timestamp as `ActiveSupport::TimeWithZone`, but Rails 7.x formats the same field as `Time`. But I didn't see this discussed in the issue, so I don't know what went wrong

**Comment by Dearest** [comment]
@ghiculescu   COOL! , I use this fix the issue
```ruby
ActiveRecord::Base.time_zone_aware_types += [:timestamptz]
```
Before add `ActiveRecord::Base.time_zone_aware_types += [:timestamptz]`
```ruby
Loading development environment (Rails 7.0.1)
[1] pry(main)> ActiveRecord::Base.time_zone_aware_attributes
=> true
[2] pry(main)> ActiveRecord::Base.time_zone_aware_types
=> [:datetime, :time]
[3] pry(main)> ActiveRecord::Base.skip_time_zone_conversion_for_attributes
=> []
```
```ruby
Loading development environment (Rails 6.1.4.1)
[1] pry(main)> ActiveRecord::Base.time_zone_aware_attributes
=> true
[2] pry(main)> ActiveRecord::Base.time_zone_aware_types
=> [:datetime, :time]
[3] pry(main)> ActiveRecord::Base.skip_time_zone_conversion_for_attributes
=> []
```

**Comment by Dearest** [comment]
@ghiculescu  But I got another issue 
in Rails6.  User.last.created_at.class is `ActiveSupport::TimeWithZone`. And I use `to_s`，I will get result like this
```ruby
Loading development environment (Rails 6.1.4.1)
[1] pry(main)> User.last.created_at.to_s
=> "2021-12-08 18:13:34"
```
But Rails7. I will get string include time zone
```ruby
Loading development environment (Rails 7.0.1)
[1] pry(main)> User.last.created_at.to_s
=> "2021-12-08 18:13:34 +0800"
```
Can I  configure the behavior of `to_s` . I don't want to use the monkey patch to hack `to_s` method

**Comment by ghiculescu** [comment]
@Dearest is `User.last.created_at.class` the same in both Rails versions?

btw I see you're on 7.0.1. If you are upgrading now, you should try going straight to [7.0.2.2](https://rubygems.org/gems/rails).

**Comment by laptopmutia** [comment]
why there is no documentation for this important config?

I think it should be mentioned here https://guides.rubyonrails.org/active_record_postgresql.html

and where to put this config?
```
ActiveRecord::ConnectionAdapters::PostgreSQLAdapter.datetime_type = :timestamptz
``` in `application.rb`?

**Comment by ghiculescu** [comment]
> I think it should be mentioned here https://guides.rubyonrails.org/active_record_postgresql.html

I agree. Would you like to make a PR?

The snippet from https://guides.rubyonrails.org/configuring.html#activerecord-connectionadapters-postgresqladapter-datetime-type should work anywhere in `config/application.rb`.

If there were no changes to `schema.rb`, then that's fine. I think that is only needed if you were manually using timestamptz columns before switching, or something like that.

**Comment by laptopmutia** [comment]
I think I still need to do the migration manually to force changes on postgresql table structure

without doing this manual migration the postgresql still gonna use `timestamp(6) without time zone` instead `timestamp(6) with time zone` for `created_at` and `updated_at`

**Comment by laptopmutia** [comment]
@ghiculescu my mistake

doing `rails db:migrate` after change config to use timestampz 
```
# config/application.rb

ActiveSupport.on_load(:active_record_postgresqladapter) do
  self.datetime_type = :timestamptz
end
```

indeed changing something on schema.rb

the changes are 

from this
```
ActiveRecord::Schema[7.1].define(version: 2024_07_31_041319) do
  # These are extensions that must be enabled in order to support this database
  enable_extension "plpgsql"

  create_table "blogs", force: :cascade do |t|
    t.string "title"
    t.text "content"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

end
```

to this 

```
  create_table "blogs", force: :cascade do |t|
    t.string "title"
    t.text "content"
    t.timestamp "created_at", precision: 6, null: false
    t.timestamp "updated_at", precision: 6, null: false
  end
 ```
 
 what I mean by not changing before is the table structure because its still using `timestamp(6) without time zone` instead timestamp(6) with time zone`

**Comment by ghiculescu** [comment]
Yes that’s intentional. It won’t rebuild your database, it just applies to new columns going forward. We need to reset the schema file to ensure you’re generating a test database correctly.

**Comment by andyatkinson** [comment]
Hi @ghiculescu and @laptopmutia, I created this Rails Guides Active Record PostgreSQL PR to document why and how to use this new configuration option. https://github.com/rails/rails/pull/53687

This was based on generating a new Rails 8, realizing this data type is not the default, then realizing I didn't know how to configure my app to use it. I found this PR, then saw the suggestion it's not document yet, and a suggestion to put it into the Rails Guides page. I picked the data types section.

Let me know what you think!

**Comment by chaadow** [comment]
Hi @ghiculescu thanks for this PR. 

If I don't add `:timestamp` to `time_zone_aware_attributes` this breaks production for existing columns : 
```ruby
ActiveSupport.on_load(:active_record_postgresqladapter) do
  self.datetime_type = :timestamptz
  
  # Had to add this
  ActiveRecord::Base.time_zone_aware_types << :timestamp
end
```

I was confused how when I set `self.datetime_type = :timestamptz` all my `created_at/updated_at` were in UTC.
After researching it was because of these lines:
https://github.com/rails/rails/blob/77191f5afb9c39aa5968219047ae69cdb2f48b14/activerecord/lib/active_record/connection_adapters/postgresql/oid/date_time.rb#L29-L31

I think technically what you wrote is correct for fresh applications (from scratch) but for existing applications, I think it's worth mentionning that `ActiveRecord::Base.time_zone_aware_types << :timestamp` needs to be added until all columns have been migrated. and since `timestamp` was transformed into `datetime` and `datetime` is time zone aware, I believe this is a breaking change

I don't know if I should create a documentation PR, or add `timestamp` to `time_zone_aware_types` because technically for all rails applicaitons, `timestamp` is equal to `datetime`
